File: pocketcloud/poc/src/crypto/encoding.js
// Fixed-point encoding of real numbers into the prime field.
//
// Reals are scaled by 2^16 and rounded. A product of two encoded values is
// therefore scaled by 2^32 — the decoder takes the scale power so callers
// can decode linear results (power 2 after one multiplication level)
// without an interactive truncation protocol, which a PoC does not need.
// Negative numbers live in the top half of the field, exactly like two's
// complement.

import { P, mod } from './field.js';

export const SCALE_BITS = 16n;
export const SCALE = 1n << SCALE_BITS;

const HALF_P = P >> 1n;

export function encode(x) {
  if (!Number.isFinite(x)) throw new Error(`cannot encode ${x}`);
  return mod(BigInt(Math.round(x * Number(SCALE))));
}

export const encodeVector = (xs) => xs.map(encode);
export const encodeMatrix = (rows) => rows.map(encodeVector);

// scalePower: how many SCALE factors the value carries.
// 1 = raw encoded value, 2 = product of two encoded values, etc.
export function decode(v, scalePower = 1) {
  const centered = v > HALF_P ? v - P : v;
  return Number(centered) / Number(SCALE ** BigInt(scalePower));
}

export const decodeVector = (vs, scalePower = 1) =>
  vs.map((v) => decode(v, scalePower));

File: pocketcloud/crypto/src/encoding.rs
//! Fixed-point encoding of reals into the field (SPEC-001 §2).
//!
//! Reals are scaled by 2^16 and rounded; negatives occupy the top half of the
//! field. A product of two encoded values carries scale power 2; the decoder
//! takes the scale power explicitly. Matches the Node PoC exactly.

use crate::field::Field;

/// Fixed-point scale bits (2^16). Matches the PoC.
pub const SCALE_BITS: u32 = 16;
/// Fixed-point scale factor.
pub const SCALE: i128 = 1i128 << SCALE_BITS;

/// Encode a real into the field: `round(x * 2^16) mod p`.
pub fn encode<F: Field>(x: f64) -> F {
    assert!(x.is_finite(), "cannot encode non-finite value");
    let scaled = (x * SCALE as f64).round() as i128;
    F::from_i128(scaled)
}

/// Decode a field element carrying `scale_power` scale factors back to a real.
pub fn decode<F: Field>(v: &F, scale_power: u32) -> f64 {
    let denom = (SCALE as f64).powi(scale_power as i32);
    v.centered() as f64 / denom
}

/// Encode a vector of reals.
pub fn encode_vec<F: Field>(xs: &[f64]) -> Vec<F> {
    xs.iter().map(|&x| encode(x)).collect()
}

/// Decode a vector at the given scale power.
pub fn decode_vec<F: Field>(vs: &[F], scale_power: u32) -> Vec<f64> {
    vs.iter().map(|v| decode(v, scale_power)).collect()
}
